allennlp/modules/seq2seq_encoders/stacked_self_attention.py

```diff
@@ -0,0 +1,159 @@
+                 non_linear_hidden_dim: int,
+                 num_layers: int,
+                 num_attention_heads: int,
+                 internal_similarity: SimilarityFunction = DotProductSimilarity(scale_output=True),
+                 use_positional_encoding: bool = True,
+                 dropout_prob: float = 0.2) -> None:
+        super(StackedSelfAttentionEncoder, self).__init__()
+
+        self._use_positional_encoding = use_positional_encoding
+        self._attention_layers: List[IntraSentenceAttentionEncoder] = []
+        self._feedfoward_layers: List[FeedForward] = []
+        self._layer_norm_layers: List[LayerNorm] = []
+
+        feedfoward_input_dim = input_dim
+        for i in range(num_layers):
+            # Project output of attention encoder through a feedforward network
+            # and back to the input size for the next layer.
+            feedfoward = FeedForward(feedfoward_input_dim,
+                                     activations=[Activation.by_name('relu')(),
+                                                  Activation.by_name('linear')()],
+                                     hidden_dims=[non_linear_hidden_dim, hidden_dim],
+                                     num_layers=2)
+
+            self.add_module(f"feedforward_{i}", feedfoward)
+            self._feedfoward_layers.append(feedfoward)
+
+            similarity_function = MultiHeadedSimilarity(num_attention_heads,
+                                                        tensor_1_dim=hidden_dim,
+                                                        internal_similarity=internal_similarity)
+            self_attention = IntraSentenceAttentionEncoder(hidden_dim,
+                                                           projection_dim,
+                                                           similarity_function,
+                                                           num_attention_heads,
+                                                           combination='1,2',
+                                                           output_projection_dim=hidden_dim)
+
+            self.add_module(f"self_attention_{i}", self_attention)
+            self._attention_layers.append(self_attention)
+
+            layer_norm = LayerNorm(self_attention.get_output_dim())
+            self.add_module(f"layer_norm_{i}", layer_norm)
+            self._layer_norm_layers.append(layer_norm)
+
+            feedfoward_input_dim = hidden_dim
+
+        self._dropout = Dropout(p=dropout_prob)
+        self._input_dim = input_dim
+        self._output_dim = self._attention_layers[-1].get_output_dim()
+
+    @overrides
+    def get_input_dim(self) -> int:
+        return self._input_dim
+
+    @overrides
+    def get_output_dim(self) -> int:
+        return self._output_dim
+
+    def forward(self, inputs: torch.Tensor, mask: torch.Tensor): # pylint: disable=arguments-differ
+        if self._use_positional_encoding:
+            output = add_positional_features(inputs)
```

Comment:
I'd prefer `inputs = add_positional_features` here, dropping the `else` clause, and using `cached_input = inputs` below.

Reply:
Hmm, nevermind, that breaks the loop.